### ISSUE TITLE
Could not initialize class org.owasp.esapi.logging.java.JavaLogFactory (ESAPI 2.2.1.0)

### ISSUE THREAD

**Issue by elangoravi** [ESAPI/esapi-java-legacy#560]
Hi Team,

We've been migrating from ESAPI 2.1.0.1 version to the latest version 2.2.1.0

When I checked the release notes, I could understand the logging implementations/structure have undergone some major changes.

All I did was include the latest version and changed the property as below 
**ESAPI.Logger=org.owasp.esapi.logging.java.JavaLogFactory**

And it throws the below exception with our existing code:
Caused by: java.lang.NoClassDefFoundError: Could not initialize class org.owasp.esapi.logging.java.JavaLogFactory
        at java.lang.Class.forName0(Native Method)
        at java.lang.Class.forName(Class.java:264)
        at org.owasp.esapi.util.ObjFactory.loadClassByStringName(ObjFactory.java:158)
        at org.owasp.esapi.util.ObjFactory.make(ObjFactory.java:81)
        at org.owasp.esapi.ESAPI.logFactory(ESAPI.java:139)
        at org.owasp.esapi.ESAPI.getLogger(ESAPI.java:155)
        at org.owasp.esapi.reference.DefaultHTTPUtilities.<init>(DefaultHTTPUtilities.java:137)
        at org.owasp.esapi.reference.DefaultHTTPUtilities.getInstance(DefaultHTTPUtilities.java:93)

(checked the release notes of 2.2.0.0 and added the necessary properties and this version is working fine with our existing code)

Apologies if I'm being naive/blunt.

Many thanks

**Comment by jeremiahjstacey** [comment]
There are 2 cases where I've seen NoClassDefFoundErrors previously.
The easy one would be if the path to the JavaLogFactory was off.  This is not the case, your definition is correct from what I'm seeing.  `org.owasp.esapi.logging.java.JavaLogFactory`
The harder one to debug is a failure during static initialization.  There are several static blocks in the class that attempt to initialize the instance.
https://github.com/ESAPI/esapi-java-legacy/blob/develop/src/main/java/org/owasp/esapi/logging/java/JavaLogFactory.java

They rely on several properties being defined in ESAPI.properties.  

The properties are defined in the DefaultSecurityConfigurationClass
https://github.com/ESAPI/esapi-java-legacy/blob/develop/src/main/java/org/owasp/esapi/reference/DefaultSecurityConfiguration.java
```
   public static final String APPLICATION_NAME = "Logger.ApplicationName";
    public static final String LOG_ENCODING_REQUIRED = "Logger.LogEncodingRequired";
    public static final String LOG_APPLICATION_NAME = "Logger.LogApplicationName";
    public static final String LOG_SERVER_IP = "Logger.LogServerIP";
    public static final String LOG_USER_INFO = "Logger.UserInfo";
    public static final String LOG_CLIENT_INFO = "Logger.ClientInfo";
```

Please verify that you have all of these values defined in your ESAPI.properties.

**Comment by jeremiahjstacey** [comment]
Another possible issue is when the class tries to load the java logging configurations.  If that fails it writes to standard error.  Also check your logs for the message 'Failed to load esapi-java-logging.properties'

**Comment by elangoravi** [comment]
Thanks @jeremiahjstacey for your inputs.

I had missed a few logging properties mentioned above and after including those, that error is resolved.

I get a null pointer exception for not including esapi-java-logging.properties in the classpath, although I expected 'Failed to load esapi-java-logging.properties' error as mentioned.

ERROR [stderr] (default task-14) java.lang.NullPointerException
ERROR [stderr] (default task-14)   at java.io.Reader.<init>(Reader.java:78)
ERROR [stderr] (default task-14)   at java.io.InputStreamReader.<init>(InputStreamReader.java:113)
ERROR [stderr] (default task-14)   at org.jboss.logmanager.PropertyConfigurator.configure(PropertyConfigurator.java:92)
ERROR [stderr] (default task-14)   at org.jboss.logmanager.LogManager.readConfiguration(LogManager.java:514)
ERROR [stderr] (default task-14)   at org.owasp.esapi.logging.java.JavaLogFactory.readLoggerConfiguration(JavaLogFactory.java:89)
ERROR [stderr] (default task-14)   at org.owasp.esapi.logging.java.JavaLogFactory.<clinit>(JavaLogFactory.java:76)

I see a problem(or misunderstanding) in readLoggerConfiguration method of JavaLogFactory.java

The expectation here is to throw an exception when "esapi-java-logging.properties" could not be read/loaded. But there is never going to be an IOException from this statement JavaLogFactory.class.getClassLoader().getResourceAsStream("esapi-java-logging.properties") , cause **getResourceAsStream returns null** if it couldn't load the file and does not thrown an IOException. The only exception can be from getClassLoader() which is a SecurityException. Therefore null is passed to the next call and Null pointer exception from there later on.

Maybe something like the below would do,

InputStream stream = JavaLogFactory.class.getClassLoader().getResourceAsStream("esapi-java-logging.properties");
if(stream!=null)
    logManager.readConfiguration(stream);
else
    System.err.print(new IOException("Failed to load esapi-java-logging.properties.");

**Comment by kwwall** [comment]
This definitely is caused the missing "esapi-java-logging.properties" file from your classpath that we forgot to drop into the ESAPI configuration jar (see https://github.com/ESAPI/esapi-java-legacy/releases/tag/esapi-2.2.1.0) and failed to provide instructions for. I will update the 2.2.1.0 release notes under the 'develop' branch to describe a more complete workaround, but ultimately, I think we are going to want to put out a 2.2.1.1 patch release that addresses this.

**Comment by kwwall** [comment]
Closed per PR #562

**Comment by premsagarbhamidipati** [comment]
After adding esapi-java-logging.properties file to my classpath resources folder, issue has been resolved. Thank you very much!
